### PR TITLE
Xuejun/remove api tensor related

### PR DESCRIPTION
**Pull request by zhaixuejun1993**
### Details:
 - Remove set_partial_shape()
 - *...*

### Tickets:
 - 106273 (100091)

### CODE REVIEW

File: src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.cpp [hunk 1]
```diff
@@ -231,6 +231,7 @@ class entryFallback : public entryBase {
             for (size_t i = 0; i < local_op->get_input_size(); i++) {
                 if (dynamic_cast<ov::opset1::Parameter*>(local_op->get_input_node_ptr(i))) {
                     local_op->get_input_tensor(i).set_partial_shape(input_shapes[i].to_partial_shape());
+                    // Question: for this "set_partial_shape" call, I think friend class is not a good way. Any other suggestions?
```
[the remark this answers is not in the document]
Comment (ilyachur) [reply src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.cpp:234]: @jane-intel Can we avoid `set_partial_shape()` call from shape inference methods? I think here we can change parameter partial shape

File: src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.cpp [hunk 2]
```diff
@@ -231,6 +231,7 @@ class entryFallback : public entryBase {
             for (size_t i = 0; i < local_op->get_input_size(); i++) {
                 if (dynamic_cast<ov::opset1::Parameter*>(local_op->get_input_node_ptr(i))) {
                     local_op->get_input_tensor(i).set_partial_shape(input_shapes[i].to_partial_shape());
```
Comment (ilyachur) [inline src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.cpp:233]: I propose to try next solution:
```suggestion
                if (auto parameter = dynamic_cast<ov::opset1::Parameter*>(local_op->get_input_node_ptr(i))) {
                    parameter.set_partial_shape(input_shapes[i].to_partial_shape());
```

Comment (zhaixuejun1993) [reply src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.cpp]: @ilyachur this suggestion doesn't work. This changes caused dynamic shape inference error, details can be found in simple PR [16230](https://github.com/openvinotoolkit/openvino/pull/16230)
I had reproduced the test error in local, get following output, 
image
**Any other suggestions?**

**Comment by wenjiew** [comment]
Should we remove [DONT REVIEW] prefix in the PR title? Thanks!

File: src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.cpp (hunk 2 above)
Comment (ilyachur) [reply src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.cpp]: @jane-intel Could you take a look? Can we avoid the usage of this internal API?

**Comment by zhaixuejun1993** [comment]
> Should we remove [DONT REVIEW] prefix in the PR title? Thanks!

Removed! Thanks!

File: src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.cpp [hunk 3]
```diff
@@ -222,13 +222,12 @@ class entryFallback : public entryBase {
             local_op = op->clone_with_new_inputs(new_inputs);
         } else {
             local_op = local_op_default;
-            OPENVINO_SUPPRESS_DEPRECATED_START
             for (size_t i = 0; i < local_op->get_input_size(); i++) {
-                if (dynamic_cast<ov::opset1::Parameter*>(local_op->get_input_node_ptr(i))) {
-                    local_op->get_input_tensor(i).set_partial_shape(input_shapes[i].to_partial_shape());
+                auto parameter = dynamic_cast<ov::opset1::Parameter*>(local_op->get_input_node_ptr(i));
+                if (parameter != nullptr) {
+                    parameter->set_partial_shape(input_shapes[i].to_partial_shape());
```
Comment (jane-intel) [inline src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.cpp:228]: after setting the shape call for `parameter->validate_and_infer_types()`. That should do the job.

Comment (jane-intel) [inline src/plugins/intel_cpu/src/utils/shape_inference/shape_inference.cpp:228]:
```suggestion
                if (auto parameter = dynamic_cast<ov::opset1::Parameter*>(local_op->get_input_node_ptr(i))) {
                    parameter->set_partial_shape(input_shapes[i].to_partial_shape());
                    parameter->validate_and_infer_types();
```

File: src/plugins/intel_cpu/tests/functional/single_layer_tests/convolution.cpp [hunk 4]
```diff
@@ -144,7 +144,10 @@ class ConvolutionLayerCPUTest : public testing::WithParamInterface<convLayerCPUT
                     }
 
                     std::vector<ov::Shape> secondParameterShapes;
-                    opToShapeInfer->get_input_tensor(0).set_partial_shape(targetShapes.front());
+                    auto parameter = dynamic_cast<ov::op::v0::Parameter*>(opToShapeInfer->get_input_node_ptr(0));
+                    if (parameter != nullptr) {
+                        parameter->set_partial_shape(targetShapes.front());
+                    }
```
Comment (jane-intel) [inline src/plugins/intel_cpu/tests/functional/single_layer_tests/convolution.cpp:150]:
```suggestion
                    if (auto parameter = dynamic_cast<ov::op::v0::Parameter*>(opToShapeInfer->get_input_node_ptr(0))) {
                        parameter->set_partial_shape(targetShapes.front());
                        parameter->validate_and_infer_types();
                    }
```

File: src/plugins/intel_cpu/tests/functional/single_layer_tests/group_convolution.cpp [hunk 5]
```diff
@@ -142,7 +142,10 @@ class GroupConvolutionLayerCPUTest : public testing::WithParamInterface<groupCon
                     }
 
                     std::vector<ov::Shape> secondParameterShapes;
-                    opToShapeInfer->get_input_tensor(0).set_partial_shape(targetShapes.front());
+                    auto parameter = dynamic_cast<ov::op::v0::Parameter*>(opToShapeInfer->get_input_node_ptr(0));
+                    if (parameter != nullptr) {
+                        parameter->set_partial_shape(targetShapes.front());
+                    }
```
Comment (jane-intel) [inline src/plugins/intel_cpu/tests/functional/single_layer_tests/group_convolution.cpp:148]:
```suggestion
                    if (auto parameter = dynamic_cast<ov::op::v0::Parameter*>(opToShapeInfer->get_input_node_ptr(0))) {
                        parameter->set_partial_shape(targetShapes.front());
                        parameter->validate_and_infer_types();
                    }
```

File: src/plugins/template/backend/int_executable.cpp [hunk 6]
```diff
@@ -83,14 +83,20 @@ class TemporaryOverrideOutputs {
         for (size_t i = 0; i < args.size(); ++i) {
             auto output = node->get_input_source_output(i);
             orig_shapes.push_back(output.get_partial_shape());
-            output.get_tensor().set_partial_shape(args[i].get_shape());
+            auto parameter = dynamic_cast<ov::op::v0::Parameter*>(node->get_input_node_ptr(i));
+            if (parameter != nullptr) {
+                parameter->set_partial_shape(args[i].get_shape());
+            }
```
Comment (jane-intel) [inline src/plugins/template/backend/int_executable.cpp:89]:
```suggestion
            if (auto parameter = dynamic_cast<ov::op::v0::Parameter*>(node->get_input_node_ptr(i))) {
                parameter->set_partial_shape(args[i].get_shape());
                parameter->validate_and_infer_types();
            }
```

File: src/plugins/template/backend/int_executable.cpp [hunk 7]
```diff
@@ -83,14 +83,20 @@ class TemporaryOverrideOutputs {
...
     }
 
     ~TemporaryOverrideOutputs() {
         for (size_t i = 0; i < orig_shapes.size(); ++i) {
             auto output = node->get_input_source_output(i);
-            output.get_tensor().set_partial_shape(orig_shapes[i]);
+            auto parameter = dynamic_cast<ov::op::v0::Parameter*>(node->get_input_node_ptr(i));
+            if (parameter != nullptr) {
+                parameter->set_partial_shape(orig_shapes[i]);
+            }
```
Comment (jane-intel) [inline src/plugins/template/backend/int_executable.cpp:99]:
```suggestion
            if (auto parameter = dynamic_cast<ov::op::v0::Parameter*>(node->get_input_node_ptr(i))) {
                parameter->set_partial_shape(orig_shapes[i]);
                parameter->validate_and_infer_types();
            }
```

**Review by jane-intel — COMMENTED** [review]
I suggest to  validate shapes so that Parameter would propagate them.

**Review by jane-intel — APPROVED** [review]

**Review by ilyachur — APPROVED** [review]

File: src/plugins/template/backend/int_executable.cpp [hunk 8]
```diff
@@ -83,14 +83,20 @@ class TemporaryOverrideOutputs {
         for (size_t i = 0; i < args.size(); ++i) {
```
Comment (ilyachur) [inline src/plugins/template/backend/int_executable.cpp:83]: @zhaixuejun1993 Please try to apply `TemporaryOverrideOutputs` only for parameters before the node cycle in the call method. I hope it should help

Comment (zhaixuejun1993) [reply src/plugins/template/backend/int_executable.cpp:83]: @ilyachur I had reproduced the CI tests failed in local, the error caused by the changes in class `TemporaryOverrideOutputs`, try u suggestions but doesn't work, maybe I'm not very familiar with the concept of `TemporaryOverrideOutputs` & `Parameter`. **Can u clarify it & show more details about the suggestion change**.
image

File: src/plugins/template/backend/int_executable.cpp [hunk 9]
```diff
@@ -83,14 +83,20 @@ class TemporaryOverrideOutputs {
...
             auto output = node->get_input_source_output(i);
             orig_shapes.push_back(output.get_partial_shape());
-            output.get_tensor().set_partial_shape(args[i].get_shape());
+            if (auto parameter = dynamic_cast<ov::op::v0::Parameter*>(node->get_input_node_ptr(i))) {
+                parameter->set_partial_shape(args[i].get_shape());
+                parameter->validate_and_infer_types();
+            }
```
Comment (jane-intel) [inline src/plugins/template/backend/int_executable.cpp:89]: I assume this doesn't work -- we need to understand what shapes are being changed.

File: src/plugins/template/backend/int_executable.cpp [hunk 10]
```diff
@@ -147,6 +153,22 @@ bool ov::runtime::interpreter::INTExecutable::call(std::vector<ov::Tensor>& outp
...
+                auto tensor = input.get_tensor_ptr();
+                op_inputs.push_back(tensor_map.at(tensor));
+            }
+            overrider_vec.push_back(TemporaryOverrideOutputs(op, op_inputs));
+        }
+    }
+
     // for each ordered op in the graph
     for (const auto& op : m_nodes) {
         if (std::dynamic_pointer_cast<ov::op::v0::Parameter>(op)) {
```
Comment (zhaixuejun1993) [inline src/plugins/template/backend/int_executable.cpp:174]: @ilyachur  The `TemporaryOverrideOutputs` not for all nodes, in here the local test skip all parameter nodes

File: src/plugins/template/backend/int_executable.cpp [hunk 11]
```diff
@@ -147,6 +153,22 @@ bool ov::runtime::interpreter::INTExecutable::call(std::vector<ov::Tensor>& outp
...
+    for (const auto& op : m_nodes) {
+        if (std::string("Parameter") == op->get_type_name()) {
+            std::vector<ov::Tensor> op_inputs;
+            for (auto input : op->inputs()) {
+                auto tensor = input.get_tensor_ptr();
+                op_inputs.push_back(tensor_map.at(tensor));
+            }
+            overrider_vec.push_back(TemporaryOverrideOutputs(op, op_inputs));
+        }
+    }
```
[the remark this answers is not in the document]
Comment (zhaixuejun1993) [reply src/plugins/template/backend/int_executable.cpp:170]: `param->output()` need index. Try following changes also failed, I think we need to take more time to understand it.
```suggestion
for (const auto& param : get_parameters()) {
        std::vector<ov::Tensor> op_inputs;
        for (size_t i = 0; i < param->get_output_size(); ++i) {
            auto tensor = param->output(i).get_tensor_ptr();
            op_inputs.push_back(tensor_map.at(tensor));
        }
        overrider_vec.push_back(TemporaryOverrideOutputs(param, op_inputs));
    }
```
error: node input index is out of range

[the remark this answers is not in the document]
Comment (ilyachur) [reply src/plugins/template/backend/int_executable.cpp:170]: So you can use index == 0. Parameter always have only one output. So I changed my suggestion.

**Comment by zhaixuejun1993** [comment]
@ilyachur With your suggestion, get failed too. With error: node input index is out of range
local trying on one test:
image

**Comment by ilyachur** [comment]
> @ilyachur With your suggestion, get failed too. With error: node input index is out of range local trying on one test: image

Maybe will it be better to use debugger in order to understand why and where the issue happens?
I didn't compile this code, but the idea is simple. You can take only parameters and change shapes only for these operations. After that validate and infer should do all work

File: src/plugins/template/backend/int_executable.cpp [hunk 12]
```diff
@@ -147,6 +153,11 @@ bool ov::runtime::interpreter::INTExecutable::call(std::vector<ov::Tensor>& outp
             results_map.emplace(output, output_count);
     }
 
+    std::vector<TemporaryOverrideOutputs> overrider_vec = {};
+    for (const auto& param : m_model->get_parameters()) {
+        overrider_vec.push_back(TemporaryOverrideOutputs(param, {tensor_map.at(param->output(0).get_tensor_ptr())}));
```
[the remark this answers is not in the document]
Comment (ilyachur) [reply src/plugins/template/backend/int_executable.cpp:158]: Could you use debugger to understand what happens

[the remark this answers is not in the document]
Comment (zhaixuejun1993) [reply src/plugins/template/backend/int_executable.cpp:158]: > Could you use debugger to understand what happens

Of cause, I'm debugging

File: src/plugins/template/backend/int_executable.cpp [hunk 13]
```diff
@@ -83,14 +83,20 @@ class TemporaryOverrideOutputs {
         for (size_t i = 0; i < args.size(); ++i) {
             auto output = node->get_input_source_output(i);
```
Comment (zhaixuejun1993) [inline src/plugins/template/backend/int_executable.cpp:84]: @ilyachur For parameter node, error happened at here, `i = 0`, get `**node input index is out of range**` ERROR. The `Parameter Node` no `get_input_source_output(0)` right?

Comment (zhaixuejun1993) [reply src/plugins/template/backend/int_executable.cpp:84]: @ilyachur  & @jane-intel  Try the following test, find only change the `Parameter node` shape can not infer the other nodes shape.
```
    std::vector<TemporaryOverrideOutputs> overrider_vec = {};
    for (const auto& param : get_parameters()) {
        // auto output = node->get_input_source_output(i);
        std::cout << "param: " << param->get_name()
                  << " default shape: " << param->get_output_partial_shape(0)
                  << " new shape: " << tensor_map.at(param->output(0).get_tensor_ptr()).get_shape() << std::endl;
        param->set_partial_shape(tensor_map.at(param->output(0).get_tensor_ptr()).get_shape());
        // parameter->set_partial_shape(orig_shapes[i]);
        param->validate_and_infer_types();
        // overrider_vec.push_back(TemporaryOverrideOutputs(param, {tensor_map.at(param->output(0).get_tensor_ptr())}));
        for (const auto& op : m_nodes) {
            std::cout << "[TEST OP] " << op->get_name() << std::endl;
            for (auto input : op->inputs()) {
                if (tensor_map.find(input.get_tensor_ptr()) == tensor_map.end())
                    std::cout << " - inputs:     " << input.get_partial_shape() << std::endl;
                else
                    std::cout << " - inputs:     " << input.get_partial_shape()
                              << "\t tensor_map: " << tensor_map.at(input.get_tensor_ptr()).get_shape() << std::endl;
            }
        }
    }
```
**This is the original graph nodes shape output:**
image
**This is the graph nodes shape after set the first `Parameter node` shape:**
image
 We can find the `ReduceMean_1442` node change shape from dynamic to static success **([?,2,?,?] ->[1, 2, 3, 4])**, but the next node `Subtract_1443` failed, the first input shape change success **([?,2,?,?] ->[1, 2, 3, 4])** but the second input failed **([?,2,1,1] ->NO CHANGE)**. The second input shape needs the output shape from previous node `ReduceMean_1442` I think, but no such info. 

**The func `validate_and_infer_types()` can only infer the current node shape not all nodes in the graph, am I right?**

Comment (ilyachur) [reply src/plugins/template/backend/int_executable.cpp:84]: Yes you need to call the model method

Comment (ilyachur) [reply src/plugins/template/backend/int_executable.cpp:84]: @jane-intel I thought that shape propagation should work in this case... Am I wrong?

Comment (ilyachur) [reply src/plugins/template/backend/int_executable.cpp:84]: @zhaixuejun1993 Please push the latest changes to this branch. It is needed in order to provide any advice.
Right now I see that situation looks like shapes weren't propagated from the parameters to other operations.

Also after `set_partial_shape()` for all parameters you MUST call `ov::Model::validate_and_infer_types()`.

File: src/plugins/template/backend/int_executable.cpp (hunk 8 above)
Comment (ilyachur) [inline src/plugins/template/backend/int_executable.cpp:83]: We can change signature to:
```
TemporaryOverrideOutputs(std::shared_ptr<ov::Parameter> param, ...)
```

File: src/plugins/template/backend/int_executable.cpp [hunk 14]
```diff
@@ -83,14 +83,20 @@ class TemporaryOverrideOutputs {
...
             orig_shapes.push_back(output.get_partial_shape());
-            output.get_tensor().set_partial_shape(args[i].get_shape());
+            if (auto parameter = dynamic_cast<ov::op::v0::Parameter*>(node->get_input_node_ptr(i))) {
```
Comment (ilyachur) [inline src/plugins/template/backend/int_executable.cpp:86]: If you change the signature we can avoid these dynamic casts

File: src/plugins/template/backend/int_executable.cpp [hunk 15]
```diff
@@ -147,6 +153,11 @@ bool ov::runtime::interpreter::INTExecutable::call(std::vector<ov::Tensor>& outp
             results_map.emplace(output, output_count);
     }
 
+    std::vector<TemporaryOverrideOutputs> overrider_vec = {};
+    for (const auto& param : m_model->get_parameters()) {
+        overrider_vec.push_back(TemporaryOverrideOutputs(param, {tensor_map.at(param->output(0).get_tensor_ptr())}));
+    }
+
```
Comment (ilyachur) [inline src/plugins/template/backend/int_executable.cpp:160]: You must call `model->validate_and_infer_types()` here

File: src/plugins/template/backend/int_executable.cpp (hunk 13 above)
Comment (zhaixuejun1993) [reply src/plugins/template/backend/int_executable.cpp:84]: The root cause is the func `validate_nodes_and_infer_types()` call.
First set all parameter node shape, call `validate_nodes_and_infer_types()` by one of the parameter node failed, but by model success.

File: src/core/src/descriptor/tensor.cpp [hunk 16]
```diff
@@ -38,17 +38,13 @@ ov::descriptor::Tensor::Tensor(const element::Type& element_type,
 OPENVINO_SUPPRESS_DEPRECATED_START
```
[the remark this answers is not in the document]
Comment (zhaixuejun1993) [reply src/core/src/descriptor/tensor.cpp:38]: This will be removed with deprecated API set_element_type().

**Comment by zhaixuejun1993** [comment]
Blocked by this ticket: https://jira.devtools.intel.com/browse/CVS-107300

**Pull request merged**